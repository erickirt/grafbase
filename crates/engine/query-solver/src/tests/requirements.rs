use crate::{assert_solving_snapshots, tests::WithExtensions};

const SCHEMA: &str = r#"
type Product
  @join__type(graph: EXT, key: "id", resolvable: false)
  @join__type(graph: EXT, key: "name", resolvable: false)
  @join__type(graph: GQL, key: "id")
  @join__type(graph: GQL2, key: "name")
{
  args: JSON @join__field(graph: EXT)
  id: ID! @join__field(graph: EXT) @join__field(graph: GQL)
  name: String! @join__field(graph: EXT) @join__field(graph: GQL2)
}

type Query
{
  productBatch(input: Lookup! @composite__is(graph: EXT, field: "{ ids: [id] } | { names: [name] }")): [Product!]! @composite__lookup(graph: EXT) @extension__directive(graph: EXT, extension: ECHO, name: "echo", arguments: {}) @join__field(graph: EXT)
  products: [Product!]! @join__field(graph: GQL)
  products2: [Product!]! @join__field(graph: GQL2)
}

enum join__Graph
{
  EXT @join__graph(name: "ext")
  GQL @join__graph(name: "gql", url: "http://127.0.0.1:33029/")
  GQL2 @join__graph(name: "gql2", url: "http://127.0.0.1:42167/")
}

enum extension__Link
{
  ECHO @extension__link(url: "file:///echo")
}

input Lookup
  @oneOf
  @join__type(graph: EXT)
{
  ids: [ID!]
  names: [String!]
}

scalar JSON
"#;

#[test]
fn intermediate_resolver() {
    let schema = WithExtensions::new(SCHEMA).resolver("file:///echo", r#"directive @echo on FIELD_DEFINITION"#);
    assert_solving_snapshots!(
        "intermediate_resolver",
        schema,
        r#"
        query { products { id args } }
        "#
    );
}
