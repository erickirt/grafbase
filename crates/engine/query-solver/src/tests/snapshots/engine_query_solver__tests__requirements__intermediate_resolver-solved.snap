---
source: crates/engine/query-solver/src/tests/requirements.rs
expression: "digraph {\n    0 [ label = \"root\", color=forestgreen ]\n    1 [ label = \"Product.args\", color=forestgreen style=bold ]\n    2 [ label = \"Lookup/echo/echo-1.0.0#ext\", shape=parallelogram, color=forestgreen ]\n    3 [ label = \"Root#gql\", shape=parallelogram, color=forestgreen ]\n    4 [ label = \"Lookup/echo/echo-1.0.0#ext\", shape=parallelogram, color=royalblue style=dashed ]\n    5 [ label = \"Product.id\", color=forestgreen style=bold ]\n    2 -> 1 [ color=forestgreen,fontcolor=forestgreen ]\n    3 -> 2 [ label = \"10\", color=forestgreen,fontcolor=forestgreen ]\n    0 -> 3 [ label = \"10\", color=forestgreen,fontcolor=forestgreen ]\n    4 -> 1 [ color=royalblue,fontcolor=royalblue,style=dashed ]\n    3 -> 4 [ label = \"10\", color=royalblue,fontcolor=royalblue,style=dashed ]\n    3 -> 5 [ color=forestgreen,fontcolor=forestgreen ]\n    4 -> 5 [ color=royalblue,fontcolor=royalblue,style=dashed ]\n}\n"
---
digraph {
    0 [ label = "root", steiner=1 ]
    1 [ label = "Product.args", steiner=1 ]
    2 [ label = "Lookup/echo/echo-1.0.0#ext", steiner=1 ]
    3 [ label = "Root#gql", steiner=1 ]
    4 [ label = "Lookup/echo/echo-1.0.0#ext", steiner=0 ]
    5 [ label = "Product.id", steiner=1 ]
    2 -> 1 [ cost=0, steiner=1]
    3 -> 2 [ cost=10, steiner=1]
    0 -> 3 [ cost=10, steiner=1]
    4 -> 1 [ cost=0, steiner=0]
    3 -> 4 [ cost=10, steiner=0]
    3 -> 5 [ cost=0, steiner=1]
    4 -> 5 [ cost=0, steiner=0]
}
